File: app/src/main/java/com/mmutert/freshfreezer/data/ItemRepository.java
package com.mmutert.freshfreezer.data;

import android.app.Application;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.Transformations;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;


public class ItemRepository {

    private final ItemDao mItemDao;
    private final LiveData<List<FrozenItem>> mAllActiveItems;
    private final LiveData<List<FrozenItem>> mAllArchivedFrozenItems;
    private final LiveData<List<ItemNotification>> mAllNotifications;


    public ItemRepository(Application app) {

        ItemDatabase database = ItemDatabase.getDatabase(app);
        mItemDao                = database.itemDao();
        LiveData<List<FrozenItem>> mAllItems = mItemDao.getAllItems();
        mAllArchivedFrozenItems = Transformations.map(mAllItems, input -> {
            ArrayList<FrozenItem> result = new ArrayList<>();
            for (FrozenItem item : input) {
                if(item.isArchived()) {
                    result.add(item);
                }
            }
            return result;
        });
        mAllActiveItems = Transformations.map(mAllItems, input -> {
            ArrayList<FrozenItem> result = new ArrayList<>();
            for (FrozenItem item : input) {
                if(!item.isArchived()) {
                    result.add(item);
                }
            }
            return result;
        });
        mAllNotifications = mItemDao.getAllNotificationsLiveData();
    }


    /**
     * Get all the items that are not marked as archived.
     * @return The active items
     */
    public LiveData<List<FrozenItem>> getAllActiveFrozenItems() {

        return mAllActiveItems;
    }


    public void insertItem(final FrozenItem itemToInsert) {

        ItemDatabase.databaseWriteExecutor.execute(() -> {
            long rowId = mItemDao.insertItem(itemToInsert);
            itemToInsert.setId(rowId);
        });
    }


    public void deleteItem(FrozenItem itemToDelete) {

        List<ItemNotification> allNotifications = getAllNotificationsLiveData(itemToDelete).getValue();

        ItemDatabase.databaseWriteExecutor.execute(() -> {
            mItemDao.deleteItem(itemToDelete);
        });

        for (ItemNotification allNotification : allNotifications) {
            deleteNotification(allNotification);
        }
    }


    public void updateItem(final FrozenItem item) {

        ItemDatabase.databaseWriteExecutor.execute(() -> mItemDao.updateFrozenItem(item));
    }


    public void archiveItem(FrozenItem itemToArchive) {

        if (!itemToArchive.isArchived()) {
            itemToArchive.setArchived(true);

            ItemDatabase.databaseWriteExecutor.execute(() -> mItemDao.updateFrozenItem(itemToArchive));
        }
    }


    public void restoreItem(FrozenItem itemToRestore) {

        if (itemToRestore.isArchived()) {
            itemToRestore.setArchived(false);

            ItemDatabase.databaseWriteExecutor.execute(() -> mItemDao.updateFrozenItem(itemToRestore));
        }
    }


    public ItemAndNotifications getItemAndNotifications(long itemId) {

        return mItemDao.getItemAndNotifications(itemId);
    }


    public LiveData<List<ItemNotification>> getNotifications() {

        return mAllNotifications;
    }


    public LiveData<List<ItemNotification>> getAllNotificationsLiveData(FrozenItem item) {

        // TODO Refactor as Transformation on the notification live data member
        return mItemDao.getAllNotificationsLiveData(item);
    }


    public List<ItemNotification> getAllNotifications(FrozenItem item) {

        return mItemDao.getAllNotifications(item);
    }


    /**
     * Add the notification to the database
     *
     * @param notification The notification to add
     */
    public void addNotification(final ItemNotification notification) {

        ItemDatabase.databaseWriteExecutor.execute(() -> {
            mItemDao.addNotification(notification);
        });
    }


    /**
     * Remove the given notification from the database
     *
     * @param notification The notification to remove
     */
    public void deleteNotification(ItemNotification notification) {

        ItemDatabase.databaseWriteExecutor.execute(() -> {
            mItemDao.deleteNotification(notification);
        });
    }


    public LiveData<List<FrozenItem>> getAllArchivedFrozenItems() {

        return mAllArchivedFrozenItems;
    }


    public LiveData<ItemAndNotifications> getItemAndNotificationsLiveData(final long itemId) {

        // TODO Refactor as Transformation on the notification live data member and create the member
        return mItemDao.getItemAndNotificationsLiveData(itemId);
    }
}

File: app/src/main/java/com/mmutert/freshfreezer/data/ItemRepository.kt
package com.mmutert.freshfreezer.data

import androidx.lifecycle.LiveData
import androidx.lifecycle.Transformations
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class ItemRepository(private val mStoredItemDao: StoredItemDao,
                     private val mNotificationDao: NotificationDao,
                     private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO) {

    /**
     * Get all the items that are not marked as archived.
     * @return The active items
     */
    private val mAllItems = mStoredItemDao.allItems
    val allActiveStorageItems: LiveData<List<StorageItem>> =
            Transformations.map(mAllItems) { input: List<StorageItem> ->
                input.filter { !it.isArchived }
            }

    val allArchivedStorageItems: LiveData<List<StorageItem>> =
            Transformations.map(mAllItems) { input: List<StorageItem> ->
                input.filter { it.isArchived }
            }
    val notifications: LiveData<List<ItemNotification>> = mNotificationDao.allNotificationsLiveData

    suspend fun insertItem(itemToInsert: StorageItem) =
        withContext(ioDispatcher) {
            val rowId = mStoredItemDao.insertItem(itemToInsert)
            itemToInsert.id = rowId
        }

    suspend fun deleteItem(itemToDelete: StorageItem) {
        val allNotifications = getAllNotificationsLiveData(itemToDelete).value!!

        withContext(ioDispatcher) {
            mStoredItemDao.deleteItem(itemToDelete)
        }
        for (allNotification in allNotifications) {
            deleteNotification(allNotification)
        }
    }

    suspend fun updateItem(item: StorageItem) =
        withContext(ioDispatcher) {
            mStoredItemDao.updateFrozenItem(item)
        }

    suspend fun archiveItem(itemToArchive: StorageItem) {
        if (!itemToArchive.isArchived) {
            itemToArchive.isArchived = true
            withContext(ioDispatcher) {
                mStoredItemDao.updateFrozenItem(itemToArchive)
            }
        }
    }

    suspend fun restoreItem(itemToRestore: StorageItem) {
        if (itemToRestore.isArchived) {
            itemToRestore.isArchived = false
            withContext(ioDispatcher) {
                mStoredItemDao.updateFrozenItem(itemToRestore)
            }
        }
    }

    suspend fun getItemAndNotifications(itemId: Long): ItemAndNotifications =
        withContext(ioDispatcher) {
            return@withContext mStoredItemDao.getItemAndNotifications(itemId)!!
        }

    fun getAllNotificationsLiveData(item: StorageItem): LiveData<List<ItemNotification>> {

        // TODO Refactor as Transformation on the notification live data member
        return mNotificationDao.getAllNotificationsLiveData(item)
    }

    fun getAllNotifications(item: StorageItem): List<ItemNotification> {
        return mNotificationDao.getAllNotifications(item)
    }

    /**
     * Add the notification to the database
     *
     * @param notification The notification to add
     */
    suspend fun addNotification(notification: ItemNotification) {
        withContext(ioDispatcher) { mNotificationDao.addNotification(notification) }
    }

    /**
     * Remove the given notification from the database
     *
     * @param notification The notification to remove
     */
    suspend fun deleteNotification(notification: ItemNotification) {
        withContext(ioDispatcher) { mNotificationDao.deleteNotification(notification) }
    }

    fun getItemAndNotificationsLiveData(itemId: Long): LiveData<ItemAndNotifications> {

        // TODO Refactor as Transformation on the notification live data member and create the member
        return mStoredItemDao.getItemAndNotificationsLiveData(itemId)
    }

    suspend fun getStorageItem(itemId: Long): StorageItem =
            // TODO Check to see if there should be some error checking before the assertion
            withContext(ioDispatcher) {
                return@withContext mStoredItemDao.getStoredItem(itemId)!!
            }
}